crates/wasmtime/src/runtime/vm/libcalls.rs

```diff
@@ -808,15 +809,113 @@ unsafe fn array_new_elem(
 
 #[cfg(feature = "gc")]
 unsafe fn array_init_elem(
-    _instance: &mut Instance,
-    _array_type_index: u32,
-    _array: u32,
-    _dst_index: u32,
-    _elem_index: u32,
-    _src: u32,
-    _len: u32,
+    instance: &mut Instance,
+    array_type_index: u32,
+    array: u32,
+    dst: u32,
+    elem_index: u32,
+    src: u32,
+    len: u32,
 ) -> Result<()> {
-    bail!("the `array.init_elem` instruction is not yet implemented")
+    use crate::{
+        store::AutoAssertNoGc,
+        vm::const_expr::{ConstEvalContext, ConstExprEvaluator},
+        ArrayRef, Func, Val,
+    };
+    use wasmtime_environ::{ModuleInternedTypeIndex, TableSegmentElements};
+
+    // NB: Don't use `OpaqueRootScope` here because we need to borrow the store
+    // through `instance` during const evaluation, which is within the same
+    // region that the `OpaqueRootScope` would otherwise span while borrowing
+    // the same store, resulting in double borrows.
+    instance.with_gc_lifo_scope(|instance| {
+        // Convert the indices into their typed forms.
+        let _array_type_index = ModuleInternedTypeIndex::from_u32(array_type_index);
```

Comment:
Should this be removed?

Reply:
I left it in just so that if we ever refactor this implementation and need to use it, we don't mix up which kind of index it is.